core/zip/src/RZip.cxx

```diff
@@ -130,24 +130,23 @@ void R__zipPrecisionCascade(int *srcsize, char *src, int *tgtsize, char **tgts,
       memset(irep,0,tgt_number*sizeof(int));
       return;
    }
-
    auto content = reinterpret_cast<ROOT::Internal::PrecisionCascadeConfigArrayContent*>(configarray);
-   (void) configsize;
    assert(content && (content->SizeOf() == (size_t)configsize));
    Int_t *cxlevels = content->GetLevels();  // This an array of size `content->fLen`
 
-   for (int tgt_idx=0; tgt_idx<tgt_number; tgt_idx++) {
-      // can only be 0 for the last of multiple (not just one) targets
-      // otherwise must be a positive value
-      int cxlevel_min = (tgt_idx > 0 && tgt_idx == tgt_number-1  ? 0 : 1);
-      if (cxlevels[tgt_idx] < cxlevel_min) {
+   // do not accept negative levels, and only accept level 0 if it is the final level,
+   // as this is a proxy for storing the residual of a cascade
+   bool storeResidual = content->fStoreResidual || (tgt_number > 0 && cxlevels[tgt_number-1] == 0);
+   int tgt_checks = tgt_number - (storeResidual ? 1 : 0);
```

Comment:
Humm that seems wrong.  If 'content->fStoreResidual', the last element is probably not zero, unless we explicitly already added it (in which case the if can be simplified to `tgt_number > 0 && cxlevels[tgt_number-1] == 0)`, isn't it?

Reply:
Thanks for double-checking me on this, Philippe.

I have observed that if 'content->fStoreResidual', then tgt_number, which is passes via precisionCascades.size() from TBasket, is 1 larger than `content->fLen', so there is no point in examining cxlevels[tgt_number-1] in that case because the user never set that many levels. In other words, if I provide 5 levels, and turn on fStoreResidual, then precisionCascades will have a size of 6 (it already has an extra buffer for the residual) and I only want to check the value of the 5 provided levels [0]...[4], not [5]. So I think this is the correct thing to do.

Reply:
but what if the user put both a zero at the end and has `fStoreResidual == true` ?

Reply:
Yeah, right now it is set to fail on that with the check on the level being less than 1.  The code you've written will create two buffers and output files for the residual in that case, so if it is to be allowed, it should be caught further upstream of this to avoid creating two buffers/files for the residual.

Reply:
Maybe the simplest is to not support '0' as a compression level (and possible explicitly reject it in `PrecisionCascadeCompressionConfig::PrecisionCascadeCompressionConfig` in `Compression.cxx`.  then the code above can be simplified, right?

Reply:
Perhaps this was the only controversial point in this PR?

Reply:
yes but since we migrate the i/o to store the 'storeResidual' at the end of the levels array (adding it based on the user input), it seems most of the PR is moot.   Let me finish the change and see what I need to import.